Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright 2023, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+/// An in-process implementation of a ``ClientTransport``.
+///
+/// This is useful when you're interested in testing your application without any actual networking layers
+/// involved, as the client and server will communicate directly with each other via in-process streams.
+///
+/// To use this client, you'll have to provide an ``InProcessServerTransport`` upon creation, as well
+/// as a ``ClientRPCExecutionConfigurationCollection``, containing a set of
+/// ``ClientRPCExecutionConfiguration``s which are specific, per-method configurations for your
+/// transport.
+///
+/// Once you have a client, you must keep a long-running task executing ``connect(lazily:)``, which
+/// will return only once all streams have been finished and ``close()`` has been called on this client; or
+/// when the containing task is cancelled.
+///
+/// To execute requests using this client, use ``withStream(descriptor:_:)``. If this function is
+/// called before ``connect(lazily:)`` is called, then any streams will remain pending and the call will
+/// block until ``connect(lazily:)`` is called or the task is cancelled.
+///
+/// - SeeAlso: ``ClientTransport``
+public struct InProcessClientTransport: ClientTransport {
+  private enum State: Sendable {
+    case unconnected(
```

Comment:
nit: prefer using state `struct`s, they're a bit better for readability (e.g. if you had two properties of the same type in an enum you could mix them up). Less important here are there aren't many states, but you can also encode valid transitions by having `init`s which take valid previous states.